src/Common/ConcurrencyControl.h

```diff
@@ -0,0 +1,234 @@
+            if (!granted)
+                return {};
+            granted--;
+            return SlotPtr(new Slot(shared_from_this()));
+        }
+
+    private:
+        friend struct Slot; // for release()
+        friend class ConcurrencyControl; // for grant(), free() and ctor
+
+        Allocation(ConcurrencyControl & parent_, SlotCount limit_, SlotCount granted_)
+            : parent(parent_)
+            , limit(limit_)
+            , allocated(granted_)
+            , granted(granted_)
+        {}
+
+        auto free()
+        {
+            std::unique_lock lock{mutex};
+            return std::pair{allocated - released,
+                allocated < limit ?
+                    std::optional<Waiters::iterator>(waiter) :
+                    std::optional<Waiters::iterator>()};
+        }
+
+        void wait(Waiters::iterator waiter_)
+        {
+            waiter = waiter_;
+        }
+
+        // Grant single slot to allocation, returns true iff more slot(s) are required
+        bool grant()
+        {
+            std::unique_lock lock{mutex};
+            granted++;
+            allocated++;
+            return allocated < limit;
+        }
+
+        // Release one slot and grant it to other allocation if required
+        void release()
+        {
+            parent.release(1);
+            std::unique_lock lock{mutex};
+            released++;
+            assert(released <= allocated);
+        }
+
+        ConcurrencyControl & parent;
+        const SlotCount limit;
+
+        std::mutex mutex; // the following values must be accessed under this mutex
+        SlotCount allocated = 0; // allocated total (including already released)
+        SlotCount granted = 0; // allocated, but not yet acquired
+        SlotCount released = 0;
+
+        Waiters::iterator waiter; // iterator to itself in Waiters list; valid iff allocated < limit
+    };
+
+public:
+    ConcurrencyControl()
+        : cur_waiter(waiters.end())
+    {}
+
+    // WARNING: all Allocation objects MUST be destructed before ConcurrencyControl
+    // NOTE: Recommended way to achieve this is to use `instance()` and do graceful shutdown of queries
+    ~ConcurrencyControl()
+    {
+        assert(waiters.empty());
+    }
+
+    // Allocate at least `min` and at most `max` slots.
+    // If not all `max` slots were successfully allocated, a subscription for later allocation is created
+    // Use Allocation::tryAcquire() to acquire allocated slot, before running a thread.
+    [[nodiscard]] AllocationPtr allocate(SlotCount min, SlotCount max)
+    {
+        std::unique_lock lock{mutex};
+
+        // Acquire as much slots as we can, but not lower than `min`
+        SlotCount limit = std::max(min, max);
+        SlotCount granted = std::max(min, std::min(limit, available(lock)));
+        cur_concurrency += granted;
+
+        // Create allocation and start waiting if more slots are required
+        auto allocation = new Allocation(*this, limit, granted);
```

Comment:
Also just because constructor is private. Using raw pointer here should be safe (unless `std::list::insert()` throws, in which case we anyways have big troubles), and it looked to me a little bit cleaner. But I can change it to shared_ptr if it's not very convincing.